common/interactivetx.c

```diff
@@ -0,0 +1,732 @@
+#include "config.h"
+#include <bitcoin/chainparams.h>
+#include <bitcoin/psbt.h>
+#include <bitcoin/script.h>
+#include <ccan/array_size/array_size.h>
+#include <ccan/cast/cast.h>
+#include <ccan/mem/mem.h>
+#include <ccan/tal/str/str.h>
+#include <common/billboard.h>
+#include <common/blockheight_states.h>
+#include <common/crypto_sync.h>
+#include <common/gossip_rcvd_filter.h>
+#include <common/gossip_store.h>
+#include <common/initial_channel.h>
+#include <common/interactivetx.h>
+#include <common/lease_rates.h>
+#include <common/memleak.h>
+#include <common/peer_billboard.h>
+#include <common/peer_failed.h>
+#include <common/psbt_internal.h>
+#include <common/psbt_open.h>
+#include <common/read_peer_msg.h>
+#include <common/setup.h>
+#include <common/status.h>
+#include <common/subdaemon.h>
+#include <common/type_to_string.h>
+#include <common/wire_error.h>
+
+/*
+ * BOLT-f53ca2301232db780843e894f55d95d512f297f9 #2:
+ * The receiving node: ...
+ * - MUST fail the negotiation if: ...
+ *  - if has received 4096 `tx_add_input` messages during this negotiation
+ */
+#define MAX_TX_ADD_INPUT_MSG_RCVD 4096
+/*
+ * BOLT-f53ca2301232db780843e894f55d95d512f297f9 #2:
+ * The receiving node: ...
+ * - MUST fail the negotiation if: ...
+ *  - it has received 4096 `tx_add_output` messages during this negotiation
+ */
+#define MAX_TX_ADD_OUTPUT_MSG_RCVD 4096
+
+/*
+ * BOLT-f53ca2301232db780843e894f55d95d512f297f9 #2:
+ * The receiving node: ...
+ * - MUST fail the negotiation if: ...
+ *  - there are more than 252 inputs
+ *  - there are more than 252 outputs
+ */
+#define MAX_FUNDING_INPUTS 252
+#define MAX_FUNDING_OUTPUTS 252
+
+struct interactivetx_context *new_interactivetx_context(const tal_t *ctx,
+							enum tx_role our_role,
+							struct per_peer_state *pps,
+							struct channel_id channel_id)
+{
+	struct interactivetx_context *ictx = tal(ctx, struct interactivetx_context);
+
+	ictx->ctx = NULL;
+	ictx->our_role = our_role;
+	ictx->pps = pps;
+	ictx->channel_id = channel_id;
+	ictx->tx_msg_count[TX_ADD_INPUT] = 0;
+	ictx->tx_msg_count[TX_ADD_OUTPUT] = 0;
+	ictx->tx_msg_count[TX_RM_INPUT] = 0;
```

Comment:
i think we can remove the "RM" counts, as the spec only counts adds (for the reason that you'll hit the add limit first!)